crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2026,6 +2028,118 @@ impl<'db> Class<'db> {
         self.iter_mro(db).contains(&ClassBase::Class(other))
     }
 
+    /// Return the explicit `metaclass` of this class, if one is defined.
+    ///
+    /// ## Note
+    /// Only call this function from queries in the same file or your
+    /// query depends on the AST of another file (bad!).
+    fn explicit_metaclass(self, db: &'db dyn Db) -> Option<Type<'db>> {
+        let class_stmt = self.node(db);
+        let metaclass_node = &class_stmt
+            .arguments
+            .as_ref()?
+            .find_keyword("metaclass")?
+            .value;
+        Some(if class_stmt.type_params.is_some() {
+            // when we have a specialized scope, we'll look up the inference
+            // within that scope
+            let model = SemanticModel::new(db, self.file(db));
+            metaclass_node.ty(&model)
+        } else {
+            // Otherwise, we can do the lookup based on the definition scope
+            let class_definition = semantic_index(db, self.file(db)).definition(class_stmt);
+            definition_expression_ty(db, class_definition, metaclass_node)
+        })
+    }
+
+    /// Return the metaclass of this class, or `Unknown` if the metaclass cannot be inferred.
+    pub(crate) fn metaclass(self, db: &'db dyn Db) -> Type<'db> {
+        // TODO: `type[Unknown]` would be a more precise fallback
+        // (needs support for <https://docs.python.org/3/library/typing.html#the-type-of-class-objects>)
+        self.try_metaclass(db).unwrap_or(Type::Unknown)
+    }
+
+    /// Return the metaclass of this class, or an error if the metaclass cannot be inferred.
+    #[salsa::tracked]
+    pub(crate) fn try_metaclass(self, db: &'db dyn Db) -> Result<Type<'db>, MetaclassError<'db>> {
+        /// Infer the metaclass of a class, tracking the classes that have been visited to detect
+        /// cyclic definitions.
+        fn infer<'db>(
+            db: &'db dyn Db,
+            class: Class<'db>,
+            seen: &mut SeenSet<Class<'db>>,
```

Comment:
Do we have tests exercising getting the metaclass of a cyclically-defined class?